### PR TITLE
drivers: flash: Generic SPI Flash driver

### PR DESCRIPTION
**Pull request by findlayfeng**
Merged #5456.
Update spi api.
Independent of spi flash config.

### CODE REVIEW

**Comment by tbursztyka** [comment]
This will need some patches split.

first patch: add your modification to get it running with latest spi API. Add your signed-off-by in the commit message

Then the second patch: add the CS/gpio logic you did, rework the Kconfig as you did etc...

Add a 3rd patch: creating the Kconfig.w25qxxdv etc..

File: drivers/flash/spi_flash_sfdp.c [hunk 1]
```diff
@@ -172,12 +180,16 @@ static inline int spi_flash_write_sr_cr(struct device *dev, u8_t sr, u8_t cr)
...
+		.buffers = &cmd,
+		.count = 1	
+	};
 
 	buf[0] = CMD_WRITE_STATUS_CONFIG;
 	buf[1] = sr;
 	buf[2] = cr;
 
-	if (spi_transceive(&data->config, &cmd, 1, &cmd, 1) != 0) {
+	if (spi_transceive(data->spi, &data->config, &cmd_set, &cmd_set) != 0) {
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: looks like spi_write() would be just fine here.

File: drivers/flash/spi_flash_sfdp.c [hunk 2]
```diff
@@ -320,10 +344,14 @@ static int spi_flash_cmd(struct device *dev, u8_t cmd_id)
...
 	};
+	struct spi_buf_set cmd_set = {
+		.buffers = &cmd,
+		.count = 1	
+	};
 
 	buf[0] = cmd_id;
 
-	if (spi_transceive(&data->config, &cmd, 1, &cmd, 1) != 0) {
+	if (spi_transceive(data->spi, &data->config, &cmd_set, &cmd_set) != 0) {
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: spi_write()

File: drivers/flash/spi_flash_sfdp.c [hunk 3]
```diff
@@ -423,7 +462,7 @@ static int spi_flash_write(struct device *dev, off_t offset,
 static int spi_flash_write_protection_set(struct device *dev, bool enable)
 {
 	struct spi_flash_data *const data = dev->driver_data;
-	int r;
+	int r = 0;
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: why? r is always assigned below in the if or else blocks.

File: drivers/flash/spi_flash_sfdp.c [hunk 4]
```diff
@@ -521,20 +564,32 @@ static const struct flash_driver_api spi_flash_api = {
 
 static int spi_flash_init(struct device *dev)
 {
-	struct device *spi_dev;
 	struct spi_flash_data *data = dev->driver_data;
-	int r;
+	int r = 0;
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: same

File: drivers/flash/spi_flash_sfdp.c [hunk 5]
```diff
@@ -548,11 +603,11 @@ static int spi_flash_init(struct device *dev)
...
+	return r;
 }
 
 static struct spi_flash_data spi_flash_memory_data;
 
-DEVICE_INIT(spi_flash_memory, "flash0", spi_flash_init,
-	    &spi_flash_memory_data, NULL, POST_KERNEL,
-	    CONFIG_SPI_FLASH_SFDP_INIT_PRIORITY);
+DEVICE_INIT(spi_flash_memory, CONFIG_SPI_FLASH_DRV_NAME,
+			spi_flash_init, &spi_flash_memory_data, NULL, POST_KERNEL,
```
Comment (tbursztyka) [inline drivers/flash/spi_flash_sfdp.c]: indentation issue: keep lines aligned to latest '(' + 1 space, as it used to be actually.

**Review by tbursztyka — CHANGES_REQUESTED** [review]

File: drivers/flash/spi_flash_sfdp.c (hunk 1 above)
Comment (findlayfeng) [reply drivers/flash/spi_flash_sfdp.c]:
```
static inline int spi_write(struct device *dev,
			    const struct spi_config *config,
			    const struct spi_buf_set *tx_bufs)
{
	return spi_transceive(dev, config, tx_bufs, NULL);
}
```

File: drivers/flash/spi_flash_sfdp.c (hunk 3 above)
Comment (findlayfeng) [reply drivers/flash/spi_flash_sfdp.c]: 这是我在更新api的时候遗留的部分修改。
我注释掉大部分代码，然后逐渐去除注释。
有可能会没有使用r。
下面是谷歌翻译的内容
Here's what Google Translates
Here are some of the changes that I left when updating the api.
I commented out most of the code and gradually removed the comments.
There may be no use of r.

File: drivers/flash/spi_flash_sfdp.c (hunk 5 above)
Comment (findlayfeng) [reply drivers/flash/spi_flash_sfdp.c]: The original file uses ‘\t’ + space.
There are different indentations in my editor. 😭

Comment (tbursztyka) [reply drivers/flash/spi_flash_sfdp.c]: try to see if your editor can be configured to apply linux code style

File: drivers/flash/spi_flash_sfdp.c (hunk 1 above)
Comment (tbursztyka) [reply drivers/flash/spi_flash_sfdp.c]: Sure, but as you can see this transaction does not care about what was read. So you can ditch the rx_bufs parameter, thus you can use spi_write directly.

**Pull request closed without merging**